velox/substrait/SubstraitToVeloxExpr.cpp

```diff
@@ -0,0 +1,267 @@
+  switch (typeCase) {
+    case ::substrait::Expression::FieldReference::ReferenceTypeCase::
+        kDirectReference: {
+      auto dRef = sField.direct_reference();
+      int32_t colIdx = subParser_->parseReferenceSegment(dRef);
+      auto fieldName = subParser_->makeNodeName(inputPlanNodeId, colIdx);
+      // TODO: Get the input type and support different types here.
+      return std::make_shared<const core::FieldAccessTypedExpr>(
+          DOUBLE(), fieldName);
+    }
+    default:
+      VELOX_NYI(
+          "Substrait conversion not supported for Reference '{}'", typeCase);
+  }
+}
+
+std::shared_ptr<const core::ITypedExpr>
+SubstraitVeloxExprConverter::toVeloxExpr(
+    const ::substrait::Expression::ScalarFunction& sFunc,
+    const int32_t& inputPlanNodeId) {
+  std::vector<std::shared_ptr<const core::ITypedExpr>> params;
+  for (const auto& sArg : sFunc.args()) {
+    auto expr = toVeloxExpr(sArg, inputPlanNodeId);
+    params.push_back(expr);
+  }
+  auto functionId = sFunc.function_reference();
+  auto veloxFunction = subParser_->findVeloxFunction(functionMap_, functionId);
+  auto subType = subParser_->parseType(sFunc.output_type());
+  auto veloxType = toVeloxType(subType->type);
+  return std::make_shared<const core::CallTypedExpr>(
+      veloxType, std::move(params), veloxFunction);
+}
+
+std::shared_ptr<const core::ConstantTypedExpr>
+SubstraitVeloxExprConverter::toVeloxExpr(
+    const ::substrait::Expression::Literal& sLit) {
+  auto typeCase = sLit.literal_type_case();
+  switch (typeCase) {
+    case ::substrait::Expression_Literal::LiteralTypeCase::kFp64: {
+      return std::make_shared<core::ConstantTypedExpr>(sLit.fp64());
+    }
+    case ::substrait::Expression_Literal::LiteralTypeCase::kBoolean: {
+      return std::make_shared<core::ConstantTypedExpr>(sLit.boolean());
+    }
+    default:
+      VELOX_NYI(
+          "Substrait conversion not supported for type case '{}'", typeCase);
+  }
+}
+
+std::shared_ptr<const core::ITypedExpr>
+SubstraitVeloxExprConverter::toVeloxExpr(
+    const ::substrait::Expression& sExpr,
+    const int32_t& inputPlanNodeId) {
+  std::shared_ptr<const core::ITypedExpr> veloxExpr;
+  auto typeCase = sExpr.rex_type_case();
+  switch (typeCase) {
+    case ::substrait::Expression::RexTypeCase::kLiteral: {
+      veloxExpr = toVeloxExpr(sExpr.literal());
+      break;
+    }
+    case ::substrait::Expression::RexTypeCase::kScalarFunction: {
+      veloxExpr = toVeloxExpr(sExpr.scalar_function(), inputPlanNodeId);
+      break;
+    }
+    case ::substrait::Expression::RexTypeCase::kSelection: {
+      veloxExpr = toVeloxExpr(sExpr.selection(), inputPlanNodeId);
+      break;
+    }
+    default:
+      VELOX_NYI(
+          "Substrait conversion not supported for Expression '{}'", typeCase);
+  }
+  return veloxExpr;
+}
+
+// TODO: Support different types here.
+class SubstraitVeloxExprConverter::FilterInfo {
+ public:
+  FilterInfo() {}
```

Comment:
nit: no need to explicitly specify it in this case

Reply:
Fixed the cases like this.